doc: libraries: Add BM_ZMS documentation (#297)
nrfconnect/sdk-nrf-bm: doc/nrf-bm/libraries/bm_zms.rst

This ATE can appear at any position of the sector.

``ID ATE`` are entries that contain a 32-bit key and describe where the data is stored, its size, and its CRC32.

``Data`` is the actual value associated with the ID-ATE.

BM_ZMS workflow
***************

The following sections describe in detail the operations performed by this library.

Registering the user callback handler
=====================================

BM_ZMS can be used with asynchronous and synchronous storage APIs.

To use it with asynchronous APIs, it is recommended to register a callback handler that will be called when an operation (write/init) has finished.
Read operations are synchronous and do not require a callback handler.

Mounting the storage system
===========================

Mounting the storage system starts by checking that the file system properties are correct, such as the sector_size and sector_count.
Then, the ``bm_zms_mount`` function must be called to make the storage ready.

To mount the file system, the following elements in the ``bm_zms_fs`` structure must be initialized:

.. code-block:: c

	struct bm_zms_fs {

> 		/** File system offset in non volatile memory **/

Reviewer: ```suggestion
		/** File system offset in non-volatile memory **/
```

Assuming that 4 sectors exist in a partition, BM_ZMS will only use 3 sectors to store key-value pairs and keep one sector empty to be able to perform GC.
The empty sector will rotate between the 4 sectors in the partition.

.. note::
   The maximum single data length that can be written at once in a sector is 64 KB.

Small data values
=================

Values smaller than or equal to 8 bytes are stored within the entry (ATE) itself, without writing data at the top of the sector.

BM_ZMS has an entry size of 16 bytes which means that the maximum available space in a partition to store data is computed in this scenario in the following way:

.. math::

   \small\frac{(NUM\_SECTORS - 1) \times (SECTOR\_SIZE - (5 \times ATE\_SIZE)) \times (DATA\_SIZE)}{ATE\_SIZE}

Where:

``NUM_SECTOR``: Total number of sectors

``SECTOR_SIZE``: Size of the sector

``ATE_SIZE``: 16 bytes

``(5 * ATE_SIZE)``: Reserved ATEs for header and delete items

``DATA_SIZE``: Size of the small data values (range from 1 to 8)

For example for 4 sectors of 1024 bytes, free space for 8-byte length data is :math:`\frac{3 \times 944 \times 8}{16} = 1416 \, \text{ bytes}`.

Large data values
=================

> Large data values ( > 8 bytes) are stored separately at the top of the sector.

Reviewer: ```suggestion
Large data values (> 8 bytes) are stored separately at the top of the sector.
```

BM_ZMS uses a cycle count mechanism that avoids emulating erase operations for these devices.
It also guarantees that every memory location is written to only once for each cycle of a sector write.

The garbage collection operation also reduces the memory cell life expectancy as it performs write operations when moving blocks from one sector to another.
To prevent the garbage collector from affecting the life expectancy of the device, it is recommended to set the size of the partition appropriately.
Its size should be the double of the maximum size of data (including headers) that could be written in the storage.

See `Available space for user data <#available-space-for-user-data-key-value-pairs>`_.

Device lifetime calculation
===========================

Storage devices, whether they are traditional flash or newer technologies like RRAM/MRAM, have a limited life expectancy which is determined by the number of times memory cells can be erased and written to.

Flash devices are erased one page at a time as part of their functional behavior (otherwise memory cells cannot be overwritten), and for storage devices that do not require an erase

> operation, memory cells can be overwritten directly.

Reviewer: There is no need to split this line

The following is a typical scenario that allows to calculate the life expectancy of a device:

Assuming that an 8-byte variable is stored using the same ID but its content changes every minute.
The partition has 4 sectors with 1024 bytes each.
Each write of the variable requires 16 bytes of storage.

> As there are 944 bytes available for ATEs for each sector, and because BM_ZMS is a fast-forward storage system, the first location of the first sector is going to be rewritten after :math:`\frac{(944 \times 4)}{16} = 236 \text{ minutes}`.

Reviewer: ```suggestion
As there are 944 bytes available for ATEs for each sector, and because BM_ZMS is a fast-forward storage system, the first location of the first sector is going to be rewritten after :math:`\frac{(944 \times 4)}{16} = 236 \text{minutes}`.
```

In addition to the normal writes, the garbage collector will move the data that is still valid from old sectors to new ones.
The same ID and a big partition size are being used and thus no data will be moved by the garbage collector in this case.

> For storage devices that can be written to 20 000 times, the storage will last about 4 720 000 minutes (~9 years).

Reviewer: @b-gent What is the preferred way to write these numbers?
```suggestion
For storage devices that can be written to 20000 times, the storage will last about 4720000 minutes (~9 years).
```

.. _lib_bm_zms:

Bare Metal Zephyr Memory Storage (BM_ZMS)
#########################################

> Bare Metal Zephyr Memory Storage is a key-value storage system that is designed to work with all types of non-volatile storage technologies.

Reviewer 2: Zephyr uses the 100 line limit rule. NCS uses one sentence - one line rule.
